# Fundamentals of Machine Learning - Exercise 9
Goal of this excercise is to complete the hands-on experience of the classification task.

## Household Prices Dataset
https://www.kaggle.com/c/house-prices-advanced-regression-techniques/data

* ... I bet that you already know the data pretty well 😅

**Important attributes description:**
* SalePrice: The property's sale price in dollars. This is the target variable that you're trying to predict.
* MSSubClass: The building class
* BldgType: Type of dwelling
* HouseStyle: Style of dwelling
* OverallQual: Overall material and finish quality
* OverallCond: Overall condition rating
* YearBuilt: Original construction date
* Heating: Type of heating
* CentralAir: Central air conditioning
* GrLivArea: Above grade (ground) living area square feet
* BedroomAbvGr: Number of bedrooms above basement level)



In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import math

from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split, StratifiedKFold, KFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, f1_score, recall_score, precision_score, confusion_matrix, auc
from sklearn.preprocessing import OrdinalEncoder

# 🎯 Our goal is to predict if the house will be sold for more than 250k USD or not
* We will use categorized price as a **Target** variable

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/rasvob/VSB-FEI-Fundamentals-of-Machine-Learning-Exercises/master/datasets/zsu_cv1_data.csv', sep=',')
df = df.loc[:, ['SalePrice','MSSubClass','BldgType','HouseStyle','OverallQual','OverallCond','YearBuilt','Heating','CentralAir','GrLivArea','BedroomAbvGr']]
df.loc[:, ['Target']] = (df.SalePrice > 250000).astype(int)
df = df.drop(['SalePrice'], axis=1)

In [ ]:
display(df.head())
df['Target'].value_counts()

MSSubClass BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0          60     1Fam     2Story            7            5       2003   
1          20     1Fam     1Story            6            8       1976   
2          60     1Fam     2Story            7            5       2001   
3          70     1Fam     2Story            7            5       1915   
4          60     1Fam     2Story            8            5       2000   

  Heating CentralAir  GrLivArea  BedroomAbvGr  Target  
0    GasA          Y       1710             3       0  
1    GasA          Y       1262             3       0  
2    GasA          Y       1786             3       0  
3    GasA          Y       1717             3       0  
4    GasA          Y       2198             4       0

Target
0    1243
1     217
Name: count, dtype: int64

# Take a look at the features
* We will need it to answer the questions

In [ ]:
df.describe()

MSSubClass  OverallQual  OverallCond    YearBuilt    GrLivArea  \
count  1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean     56.897260     6.099315     5.575342  1971.267808  1515.463699   
std      42.300571     1.382997     1.112799    30.202904   525.480383   
min      20.000000     1.000000     1.000000  1872.000000   334.000000   
25%      20.000000     5.000000     5.000000  1954.000000  1129.500000   
50%      50.000000     6.000000     5.000000  1973.000000  1464.000000   
75%      70.000000     7.000000     6.000000  2000.000000  1776.750000   
max     190.000000    10.000000     9.000000  2010.000000  5642.000000   

       BedroomAbvGr       Target  
count   1460.000000  1460.000000  
mean       2.866438     0.148630  
std        0.815778     0.355845  
min        0.000000     0.000000  
25%        2.000000     0.000000  
50%        3.000000     0.000000  
75%        3.000000     0.000000  
max        8.000000     1.000000

## Categorial features EDA

In [ ]:
df.describe(exclude=np.number)

BldgType HouseStyle Heating CentralAir
count      1460       1460    1460       1460
unique        5          8       6          2
top        1Fam     1Story    GasA          Y
freq       1220        726    1428       1365

### BldgType

In [ ]:
df.BldgType.value_counts()

BldgType
1Fam      1220
TwnhsE     114
Duplex      52
Twnhs       43
2fmCon      31
Name: count, dtype: int64

### HouseStyle

In [ ]:
df.HouseStyle.value_counts()

HouseStyle
1Story    726
2Story    445
1.5Fin    154
SLvl       65
SFoyer     37
1.5Unf     14
2.5Unf     11
2.5Fin      8
Name: count, dtype: int64

### Heating

In [ ]:
df.Heating.value_counts()

Heating
GasA     1428
GasW       18
Grav        7
Wall        4
OthW        2
Floor       1
Name: count, dtype: int64

## Missing values

In [ ]:
df.isna().sum()

MSSubClass      0
BldgType        0
HouseStyle      0
OverallQual     0
OverallCond     0
YearBuilt       0
Heating         0
CentralAir      0
GrLivArea       0
BedroomAbvGr    0
Target          0
dtype: int64

## Labels distribution

In [ ]:
df.Target.value_counts()

Target
0    1243
1     217
Name: count, dtype: int64

# ✅ Task (2p)
Complete the following tasks:

1. 📈 Describe what operations you are performing for each of the features
    * Mainly focus on categorical features
      
2. 📌 Answer the following questions:
    * **How many values are missing?**
    * **How many instances do you have in each of the classes?**
    * 🔎 **Which metric score do you propose for the classification model performance evaluation?**
        * 💡 This depends on your previous answer
          
3. ⚡Finish your preprocessing pipeline and split the data into the Input and Output part (i.e. `X` and `y` variables)

4. 🌳 Start with the **Decision Tree**
    * Use 5-fold cross validation
    * 🔎 Will you use *standard* cross validation or *stratified* cross validation?
    * Compute mean of the obtained score values
      
5. 🚀 Select one other algorithm from https://scikit-learn.org/stable/supervised_learning.html
    * Repeat the 5-fold CV
      
6. 📒 **Write down which default model is better**

7. 📊 Experiment with hyper-parameters
    * Select at least one important parameter for the model
    * Set the parameter value range
        * You can use random values, interval of values, ...
    * Do the 5-fold CV
        * Compute mean of the obtained score values
    * Document the experiment results using tables and/or plots
    * Describe the results in a Markdown cell

8. 📒 **Write down which model (default or tuned) is the best and why**

* **Document everything you do in a Markdown cells**
    * ❌ Results interpretation figured in real-time during task check is not allowed! ❌



In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import math

from sklearn.model_selection import train_test_split, StratifiedKFold, KFold
from sklearn.metrics import accuracy_score, f1_score, recall_score, precision_score, confusion_matrix, auc
from sklearn.preprocessing import OrdinalEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier


# load data
df = pd.read_csv('https://raw.githubusercontent.com/rasvob/VSB-FEI-Fundamentals-of-Machine-Learning-Exercises/master/datasets/zsu_cv1_data.csv', sep=',')

# select features
df = df.loc[:, ['SalePrice','MSSubClass','BldgType','HouseStyle','OverallQual','OverallCond','YearBuilt','Heating','CentralAir','GrLivArea','BedroomAbvGr']]

# encode sale price as binary target
df.loc[:, ['Target']] = (df.SalePrice > 250000).astype(int)
df = df.drop(['SalePrice'], axis=1)

display(df.shape)
display(df.head())

(1460, 11)

MSSubClass BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0          60     1Fam     2Story            7            5       2003   
1          20     1Fam     1Story            6            8       1976   
2          60     1Fam     2Story            7            5       2001   
3          70     1Fam     2Story            7            5       1915   
4          60     1Fam     2Story            8            5       2000   

  Heating CentralAir  GrLivArea  BedroomAbvGr  Target  
0    GasA          Y       1710             3       0  
1    GasA          Y       1262             3       0  
2    GasA          Y       1786             3       0  
3    GasA          Y       1717             3       0  
4    GasA          Y       2198             4       0

In [ ]:
# Taking a look at the categorical data

display(df.describe())

print()
print("Categorical features")
display(df.describe(exclude=np.number))

print()
print("BldgType")
display(df.BldgType.value_counts())

print()
print("HouseStyle")
display(df.HouseStyle.value_counts())

print()
print("Heating")
display(df.Heating.value_counts())

print()
print("Missing values")
display(df.isna().sum())

print()
print("Target")
display(df.Target.value_counts())

MSSubClass  OverallQual  OverallCond    YearBuilt    GrLivArea  \
count  1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean     56.897260     6.099315     5.575342  1971.267808  1515.463699   
std      42.300571     1.382997     1.112799    30.202904   525.480383   
min      20.000000     1.000000     1.000000  1872.000000   334.000000   
25%      20.000000     5.000000     5.000000  1954.000000  1129.500000   
50%      50.000000     6.000000     5.000000  1973.000000  1464.000000   
75%      70.000000     7.000000     6.000000  2000.000000  1776.750000   
max     190.000000    10.000000     9.000000  2010.000000  5642.000000   

       BedroomAbvGr       Target  
count   1460.000000  1460.000000  
mean       2.866438     0.148630  
std        0.815778     0.355845  
min        0.000000     0.000000  
25%        2.000000     0.000000  
50%        3.000000     0.000000  
75%        3.000000     0.000000  
max        8.000000     1.000000


Categorical features


BldgType HouseStyle Heating CentralAir
count      1460       1460    1460       1460
unique        5          8       6          2
top        1Fam     1Story    GasA          Y
freq       1220        726    1428       1365


BldgType


BldgType
1Fam      1220
TwnhsE     114
Duplex      52
Twnhs       43
2fmCon      31
Name: count, dtype: int64


HouseStyle


HouseStyle
1Story    726
2Story    445
1.5Fin    154
SLvl       65
SFoyer     37
1.5Unf     14
2.5Unf     11
2.5Fin      8
Name: count, dtype: int64


Heating


Heating
GasA     1428
GasW       18
Grav        7
Wall        4
OthW        2
Floor       1
Name: count, dtype: int64


Missing values


MSSubClass      0
BldgType        0
HouseStyle      0
OverallQual     0
OverallCond     0
YearBuilt       0
Heating         0
CentralAir      0
GrLivArea       0
BedroomAbvGr    0
Target          0
dtype: int64


Target


Target
0    1243
1     217
Name: count, dtype: int64

# Encode the categorical data

- We can either encode the categorical features with an OrdinalEncoder or OneHotEncoder
- OrdinalEncoder is good for categories with an ordering
- OneHotEncoder is good for categories without a clear ordering
---
- Therefore I choose one hot encoding, because non of the categorical features have a clear ordering

In [ ]:
# Making a convenience function to encode columns

def encode_columns_one_hot(df: pd.DataFrame, columns: list[str]):
    for column in columns:
        encoded_column = pd.get_dummies(df[column], prefix=column, dtype=int)
        df = pd.concat([df, encoded_column], axis=1)
        df = df.drop(column, axis=1)
    return df

def encode_columns_ordinal(df: pd.DataFrame, columns: list[str]):
    for column in columns:
        categories = list(df[column].unique())

        encoder = OrdinalEncoder(categories=[categories])
        encoded_column = encoder.fit_transform(df[[column]])

        df.loc[:, column] = encoded_column
    return df

In [ ]:

# Encoding all categorical features with via new function

# extract the names of the categorical features
categorical_feature_names = list(df.describe(exclude=np.number).columns)

# encode the features
df = encode_columns_ordinal(df, categorical_feature_names)

# display the results
print(f"{categorical_feature_names = }")

display(df.shape)
display(df.head())


categorical_feature_names = ['BldgType', 'HouseStyle', 'Heating', 'CentralAir']


(1460, 11)

MSSubClass BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0          60      0.0        0.0            7            5       2003   
1          20      0.0        1.0            6            8       1976   
2          60      0.0        0.0            7            5       2001   
3          70      0.0        0.0            7            5       1915   
4          60      0.0        0.0            8            5       2000   

  Heating CentralAir  GrLivArea  BedroomAbvGr  Target  
0     0.0        0.0       1710             3       0  
1     0.0        0.0       1262             3       0  
2     0.0        0.0       1786             3       0  
3     0.0        0.0       1717             3       0  
4     0.0        0.0       2198             4       0

In [ ]:
# Splitting the data and the target
X, y = df.loc[:, df.columns != 'Target'], df.loc[:, 'Target']


# data scaling
# from sklearn.preprocessing import StandardScaler
# scaler = StandardScaler()
# print(scaler.fit(X, y))


# Splitting to training and testing subset
X_train, X_test, y_train, y_test = train_test_split(X, y)

print("Training subset")
print()
print(f"{X_train.shape = }")
print(f"{y_train.shape = }")

print()
print('-' * 50)
print()

print("Test subset")
print()
print(f"{X_test.shape = }")
print(f"{y_test.shape = }")

StandardScaler()
Training subset

X_train.shape = (1095, 10)
y_train.shape = (1095,)

--------------------------------------------------

Test subset

X_test.shape = (365, 10)
y_test.shape = (365,)


In [ ]:
# Train on the data with some chosen model

# Start out with the DecisionTreeClassifier

# clf = RandomForestClassifier(random_state=13)
clf = DecisionTreeClassifier(random_state=13)
clf.fit(X_train, y_train)

y_pred = clf.predict(X_test)

print("Confusion matrix")
display(confusion_matrix(y_true = y_test, y_pred = y_pred))

print(f"{f1_score(y_test, y_pred) = }")
print(f"{accuracy_score(y_test, y_pred) = }")
print(f"{precision_score(y_test, y_pred) = }")
print(f"{recall_score(y_test, y_pred) = }")

Confusion matrix


array([[294,  23],
       [ 10,  38]])

f1_score(y_test, y_pred) = 0.6972477064220184
accuracy_score(y_test, y_pred) = 0.9095890410958904
precision_score(y_test, y_pred) = 0.6229508196721312
recall_score(y_test, y_pred) = 0.7916666666666666


# Choosing a splitter

- KFold is good for balanced data
- StratifiedKFold is good when data is imbalanced


In [ ]:
df.Target.value_counts()

Target
0    1243
1     217
Name: count, dtype: int64

Our data is very imbalanced, so we'll use StratifiedKFold.

In [ ]:
kfold = StratifiedKFold(n_splits=5)
splits = kfold.split(X_train, y_train)
score = []

for train_index, test_index in splits:
    X_train_fold, X_test_fold = X_train.iloc[train_index], X_train.iloc[test_index]
    y_train_fold, y_test_fold = y_train.iloc[train_index], y_train.iloc[test_index]

    clf = DecisionTreeClassifier(random_state=13)
    clf.fit(X_train_fold, y_train_fold)

    y_pred_fold = clf.predict(X_test_fold)
    score.append(f1_score(y_test_fold, y_pred_fold))

np.mean(score), np.min(score), np.max(score)

(np.float64(0.6991140970890248),
 np.float64(0.6470588235294118),
 np.float64(0.7692307692307693))

I wrote a generic function that makes training the splits less of an eye sore.

In [ ]:
def train_splits(model, model_selection, X_data, y_data):
    """
    model = training model
    model_selection = kfolder
    X = source
    y = target
    """

    splits = model_selection.split(X_data, y_data)
    score = []

    for train_index, test_index in splits:
        X_train_fold, X_test_fold = X_data.iloc[train_index], X_data.iloc[test_index]
        y_train_fold, y_test_fold = y_data.iloc[train_index], y_data.iloc[test_index]

        model.fit(X_train_fold, y_train_fold)
        y_pred_fold = model.predict(X_test_fold)

        score.append(f1_score(y_test_fold, y_pred_fold))

    return { 'mean': np.mean(score), 'min': np.min(score), 'max': np.max(score) }

In [ ]:
default_models = [
    DecisionTreeClassifier(random_state=0),
    RandomForestClassifier(random_state=0),
]

results = []

for model in default_models:

    result = train_splits(
        model           = model,
        model_selection = StratifiedKFold(n_splits=5),
        X_data          = X_train,
        y_data          = y_train
    )

    y_pred = model.predict(X_test)
    score = f1_score(y_test, y_pred)

    results.append({
        'model': model,
        'result': result,
        'score': f1_score(y_test, y_pred),
    })

for result in results:
    display(result)

{'model': DecisionTreeClassifier(random_state=0),
 'result': {'mean': np.float64(0.7139037433155082),
  'min': np.float64(0.6666666666666666),
  'max': np.float64(0.8)},
 'score': 0.6857142857142857}

{'model': RandomForestClassifier(random_state=0),
 'result': {'mean': np.float64(0.7790481586205036),
  'min': np.float64(0.7540983606557377),
  'max': np.float64(0.8115942028985508)},
 'score': 0.7422680412371134}

# Hyper-parameter Optimization

In [ ]:
from sklearn.model_selection import GridSearchCV

# ML algorithms
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

# hyper-parameter optimizers
from sklearn.experimental import enable_halving_search_cv  # noqa
from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import HalvingRandomSearchCV
from sklearn.model_selection import HalvingGridSearchCV

# tools
from scipy.stats import uniform

Writing a convenience function.

In [ ]:
def find_optimal_classifier(
    hyper_param_searcher_klass,
    classifier_model_klass,
    cross_validator,
    hyper_param_distributions,
    X, y,
):
    classifier_model = classifier_model_klass(random_state=0)
    hyper_param_searcher = hyper_param_searcher_klass(
        estimator   = classifier_model,
        param_grid  = hyper_param_distributions,
        cv          = cross_validator,
        scoring     = 'f1'
    )

    optimal_classifier = hyper_param_searcher.fit(X, y)
    return optimal_classifier


def search_optimal_classifiers(
    searcher_klass,
    cross_validator,
    model_hyper_parameters,
):
    optimal_classifiers = []

    for model, hyper_param_distributions in model_hyper_parameters.items():
        search = find_optimal_classifier(
            hyper_param_searcher_klass  = searcher_klass,
            classifier_model_klass      = model,
            cross_validator             = cross_validator,
            hyper_param_distributions   = hyper_param_distributions,
            X                           = X_train,
            y                           = y_train,
        )

        df_result = pd.DataFrame(search.cv_results_)
        df_result = df_result[['rank_test_score', 'mean_test_score'] + [f'split{i}_test_score' for i in range(0, cross_validator.n_splits)]]
        df_result = df_result.sort_values(by='rank_test_score')

        optimal_classifiers.append({
            'estimator': search.best_estimator_,
            'results': df_result,
        })

    return optimal_classifiers

Define the search space for the models that will be used in the search for the best classifiers.

Search for the optimal classifiers - train them on the training set.

Then try out the trained model on test data.

In [ ]:
# defining models and their hyper parameter grid
# for the hyper-parameter optimizers to optimize

model_hyper_parameters = {
    DecisionTreeClassifier: {
        'min_samples_split' : range(2,20),
        'max_depth'         : range(5,10),
    },
    RandomForestClassifier: {
        'n_estimators'      : range(10,150,10),
        'max_depth'         : range(5,10),
    },
}

# learning on training dataset

optimal_classifiers = search_optimal_classifiers(
    searcher_klass          = GridSearchCV,
    cross_validator         = StratifiedKFold(n_splits=5),
    model_hyper_parameters  = model_hyper_parameters,
)

# displaying the results

for classifier in optimal_classifiers:
    display(classifier['results'])
    display(classifier['estimator'])

rank_test_score  mean_test_score  split0_test_score  split1_test_score  \
51                1         0.756765           0.738462           0.805970   
49                2         0.753915           0.705882           0.805970   
33                3         0.752923           0.727273           0.805970   
45                4         0.752742           0.716418           0.805970   
53                5         0.752296           0.685714           0.825397   
..              ...              ...                ...                ...   
73               86         0.713667           0.575758           0.812500   
58               87         0.713167           0.606061           0.800000   
74               88         0.711740           0.588235           0.812500   
56               89         0.706241           0.571429           0.800000   
54               90         0.705829           0.625000           0.764706   

    split2_test_score  split3_test_score  split4_test_score  
51           0.754098           0.750000           0.735294  
49           0.754098           0.750000           0.753623  
33           0.754098           0.750000           0.727273  
45           0.745763           0.741935           0.753623  
53           0.754098           0.750000           0.746269  
..                ...                ...                ...  
73           0.745763           0.709677           0.724638  
58           0.733333           0.718750           0.707692  
74           0.733333           0.718750           0.705882  
56           0.733333           0.718750           0.707692  
54           0.733333           0.698413           0.707692  

[90 rows x 7 columns]

DecisionTreeClassifier(max_depth=7, min_samples_split=17, random_state=0)

rank_test_score  mean_test_score  split0_test_score  split1_test_score  \
69                1         0.796644           0.793651           0.806452   
34                2         0.796349           0.769231           0.819672   
33                2         0.796349           0.769231           0.819672   
41                2         0.796349           0.781250           0.819672   
18                5         0.796011           0.787879           0.819672   
..              ...              ...                ...                ...   
43               66         0.768604           0.757576           0.812500   
42               67         0.767956           0.750000           0.781250   
2                68         0.766377           0.764706           0.806452   
44               69         0.765362           0.746269           0.781250   
1                70         0.763106           0.753623           0.806452   

    split2_test_score  split3_test_score  split4_test_score  
69           0.766667           0.787879           0.828571  
34           0.800000           0.781250           0.811594  
33           0.800000           0.781250           0.811594  
41           0.800000           0.769231           0.811594  
18           0.779661           0.781250           0.811594  
..                ...                ...                ...  
43           0.766667           0.730159           0.776119  
42           0.721311           0.781250           0.805970  
2            0.754098           0.718750           0.787879  
44           0.745763           0.741935           0.811594  
1            0.754098           0.718750           0.782609  

[70 rows x 7 columns]

RandomForestClassifier(max_depth=9, n_estimators=140, random_state=0)

We only trained on the training data. Now try out the trained model on test data.

In [ ]:
# predicting test dataset
scores = []

for classifier in optimal_classifiers:
    estimator = classifier['estimator']
    y_pred = estimator.predict(X_test)

    scores.append({
        'estimator': estimator,
        'score': f1_score(y_test, y_pred),
    })

scores

[{'estimator': DecisionTreeClassifier(max_depth=6, min_samples_split=7, random_state=0),
  'score': 0.7377049180327869},
 {'estimator': RandomForestClassifier(max_depth=8, n_estimators=60, random_state=0),
  'score': 0.75}]

The previous scores we got using **default settings**:

```python
{'model': DecisionTreeClassifier(random_state=0),
 'result': {'mean': np.float64(0.7139037433155082),
  'min': np.float64(0.6666666666666666),
  'max': np.float64(0.8)},
 'score': 0.6857142857142857}

{'model': RandomForestClassifier(random_state=0),
 'result': {'mean': np.float64(0.7790481586205036),
  'min': np.float64(0.7540983606557377),
  'max': np.float64(0.8115942028985508)},
 'score': 0.7422680412371134}
```

After hyper-parameter optimization we got:

```python
[{'estimator': DecisionTreeClassifier(max_depth=6, min_samples_split=16, random_state=0),
  'score': 0.7413793103448276},
 {'estimator': RandomForestClassifier(max_depth=6, n_estimators=50, random_state=0),
  'score': 0.8484848484848485}]
```

|                        | Default Settings   | Hyper-Param Optimized |
|------------------------|--------------------|-----------------------|
| DecisionTreeClassifer  | 0.68 | 0.74 |
| RandomForestClassifier | 0.74 | 0.85 |
